tools/crm_rule.c

```diff
@@ -0,0 +1,296 @@
+     * the CIB manager or doing anything else important.
+     */
+    if (rule_mode == crm_rule_mode_check) {
+        if (rule_id == NULL) {
+            CMD_ERR("--check requires use of --rule=\n");
+            crm_help(flag, CRM_EX_USAGE);
+        }
+    }
+
+    /* Set up some defaults. */
+    if (rule_date == NULL) {
+        rule_date = crm_time_new(NULL);
+    }
+
+    /* Where does the XML come from?  If one of various command line options were
+     * given, use those.  Otherwise, connect to the CIB and use that.
+     */
+    if (input_file != NULL) {
+        struct stat sb;
+
+        if (stat(input_file, &sb) == -1) {
+            CMD_ERR("cannot read input file %s: %s", input_file, strerror(errno));
+            exit_code = crm_errno2exit(errno);
+            goto bail;
+        }
+
+        input = filename2xml(input_file);
+
+        if (input == NULL) {
+            fprintf(stderr, "Couldn't parse input file: %s\n", input_file);
+            exit_code = CRM_EX_DATAERR;
+            goto bail;
+        }
+    } else if (safe_str_eq(input_xml, "-")) {
+        input = stdin2xml();
+
+        if (input == NULL) {
+            fprintf(stderr, "Couldn't parse input from STDIN\n");
+            exit_code = CRM_EX_DATAERR;
+            goto bail;
+        }
+    } else if (input_xml != NULL) {
+        input = string2xml(input_xml);
+
+        if (input == NULL) {
+            fprintf(stderr, "Couldn't parse input string: %s\n", input_xml);
+            exit_code = CRM_EX_DATAERR;
+            goto bail;
+        }
+    } else {
+        /* Establish a connection to the CIB manager */
+        cib_conn = cib_new();
+        rc = cib_conn->cmds->signon(cib_conn, crm_system_name, cib_command);
+        if (rc != pcmk_ok) {
+            CMD_ERR("Error connecting to the CIB manager: %s", pcmk_strerror(rc));
+            exit_code = crm_errno2exit(rc);
+            goto bail;
+        }
+    }
+
+    /* Populate working set from CIB query */
+    if (input == NULL) {
+        rc = cib_conn->cmds->query(cib_conn, NULL, &input, cib_scope_local | cib_sync_call);
+        if (rc != pcmk_ok) {
+            return -rc;
+        }
+    }
+
+    /* Populate the working set instance */
+    data_set = pe_new_working_set();
+    if (data_set == NULL) {
+        rc = -ENOMEM;
+        goto bail;
+    }
+
+    data_set->input = input;
+    data_set->now = rule_date;
+
+    /* Unpack everything. */
+    cluster_status(data_set);
+
+    /* Now do whichever operation mode was asked for.  There's only one at the
+     * moment so this looks a little silly, but I expect there will be more
+     * modes in the future.
+     */
+    switch(rule_mode) {
+        case crm_rule_mode_check:
+            rc = crm_rule_check(data_set, rule_id, rule_date);
+
+            if (rc < 0) {
+                CMD_ERR("Error checking rule: %s", pcmk_strerror(-rc));
+                exit_code = crm_errno2exit(-rc);
```

Comment:
FYI pcmk_strerror() and crm_errno2exit() both start by taking abs(rc), so it doesn't matter whether you use rc or -rc here